### DISCUSSION TITLE
CORS and Bad request when updating from 0.13 to 1.0.7

### DISCUSSION THREAD

**Discussion by slamer59** [dagster-io/dagster#9592] [Q&A]
Hello,
I deployed dagster / dagit and updated it.
With the older version 0.13 I don't have any problem to call graphql endpoint with apollo and now i am facing an issue. 

I have two origin :
- **https://dagit.deployed/graphl**
- **https://apolloclient.corp**

With the legacy : 0.13
```
export const client = new ApolloClient({
  uri: "https://dagit.deployed/graphl",
  cache: new InMemoryCache(),
});
```

was working fine. 

after update I had a CORS issue with the above code.

So I tried: 
```
export const client = new ApolloClient({
  cache: new InMemoryCache(),
  link: new HttpLink({
      uri: "https://dagit.deployed/graphl",
      fetchOptions: {
        mode: 'no-cors'
      }
  }),
});
```
And I end with: 
**createHttpLink.ts:156          POST https://dagit.deployed/graphql 400 (Bad Request)**

My graphql is fine when I use it in graphql console directly. 

I saw that from 0.13 to 1.0.7 the server is Flask to Starlette/FastAPI. 

Is it related ? 
How to come over it ? 
I tried to add [ALLOWED_HOST ](https://www.starlette.io/middleware/#trustedhostmiddleware) in env variable without sucess. 

Any tips ?

**Comment by alangenfeld** [comment]
Can you set the `--log-level` flag on `dagit` to `debug` or `trace` to try to get more information about why the request is being denied and to confirm its making it to the websever?

**Reply by slamer59** [reply]
I will check once I am back at work. In a week. Thanks for this proposal !

**Comment by simx-bb** [comment]
I am really a novice, so please be critical with what I'll write bellow :-)
I had also a CORS issue using 2 several servers.
I finally got it working by modifying 2 files in dagit so that the CORS preflight is satisfied (QUICK AND DIRTY!):
- webserver.py:
1. DagitWebserver > build_routes: 
		added "OPTIONS" to the methods:
```
def build_routes(self):
routes = (
					[
						Route("/dagit_info", self.dagit_info_endpoint),
						Route(
							"/graphql",
							self.graphql_http_endpoint,
							name="graphql-http",
							methods=["GET", "POST", "OPTIONS"],
						),
```
2. DagsterTracedCounterMiddleware > __call__ > send_wrapper:
		added "Access-Control-Allow-Origin" and "Access-Control-Allow-Headers" headers:

```
def send_wrapper(message: Message):
            if message["type"] == "http.response.start":
                counter = traced_counter.get()
                if counter and isinstance(counter, Counter):
                    headers = MutableHeaders(scope=message)
                    headers.append("x-dagster-call-counts", json.dumps(counter.counts()))
                    headers.append("Access-Control-Allow-Origin", "*")
                    if scope["method"] == "OPTIONS": # you may skip that and always set the header
                        headers.append("Access-Control-Allow-Headers", "access-control-allow-origin,content-type") # you may allow different headers (I just included what our preflight expect)
```
- graphql.py > GraphQLServer > graphql_http_endpoint:
	added the request.method == "OPTIONS" case:
```
                elif request.method == "POST":
                            content_type = request.headers.get("Content-Type", "")
                            if "application/json" in content_type:
                                try:
                                    data = await request.json()
                                except json.JSONDecodeError:
                                    body = await request.body()
                                    return PlainTextResponse(
                                        f"GraphQL request is invalid JSON:\n{body.decode()}",
                                        status_code=status.HTTP_400_BAD_REQUEST,
                                    )
                            elif "application/graphql" in content_type:
                                body = await request.body()
                                text = body.decode()
                                data = {"query": text}
                            else:
                                data = request.query_params
                        elif request.method == "OPTIONS":
                            return PlainTextResponse("Dummy: accept OPTIONS method for CORS preflight") # you'd better check that the preflight request is satisfied instead of always returning that everything is fine ;-)!
                        else:
                            return PlainTextResponse(
                                "Method Not Allowed", status_code=status.HTTP_405_METHOD_NOT_ALLOWED
                            )
```
Hope it helps.

@alangenfeld: any chance external graphql queries are accepted by dagit (with CORS) or is it out of scope? Thanks!
